controllers/slice/slice.go

```diff
@@ -0,0 +1,507 @@
+	// existingAllowedNsList = list of all namespaces that have kubeslice label
+	existingAllowedNsList := &corev1.NamespaceList{}
+	listOpts := []client.ListOption{
+		client.HasLabels([]string{AllowedNamespaceSelectorLabelKey}),
+	}
+	err := r.List(ctx, existingAllowedNsList, listOpts...)
+	if err != nil {
+		log.Error(err, "Failed to list namespaces")
+		return err
+	}
+	log.Info("reconciling", "existingAllowedNsList", existingAllowedNsList)
+	// Convert the list into a map for faster lookups. Will come in handy when we compare
+	// existing namespaces against configured namespaces.
+	type nsMarker struct {
+		ns     *corev1.Namespace
+		marked bool
+	}
+	existingAllowedNsMap := make(map[string]*nsMarker)
+
+	for _, existingAllowedNsObj := range existingAllowedNsList.Items {
+		existingAllowedNsMap[existingAllowedNsObj.ObjectMeta.Name] = &nsMarker{
+			ns:     &existingAllowedNsObj,
+			marked: false,
+		}
+	}
+	// Compare the existing list with the configured list.
+	// If a namespace is not found in the existing list, consider it as an addition event and
+	// label the namespace.
+	// If a namespace is found in the existing list, mark it to indicate that it has been verified
+	// to be valid as it is present in the configured list as well.
+	labeledAllowedNsList := []string{}
+	statusChanged := false
+	for _, cfgAllowedNs := range cfgAllowedNsList {
+		// the cfgAllowedNs should be labeled with "kubeslice.io/namespace" and we guard it against someone accidentally changing the value
+		if v, exists := existingAllowedNsMap[cfgAllowedNs]; exists && v.ns.GetLabels()[AllowedNamespaceSelectorLabelKey] == cfgAllowedNs {
+			existingAllowedNsMap[cfgAllowedNs].marked = true
+			labeledAllowedNsList = append(labeledAllowedNsList, cfgAllowedNs)
+			continue
+		}
+		// label does not exists on namespace
+		namespace := &corev1.Namespace{}
+		err := r.Get(ctx, types.NamespacedName{Name: cfgAllowedNs}, namespace)
+		if err != nil {
+			log.Info("Failed to find namespace", "namespace", cfgAllowedNs)
+			continue
+		}
+		// A namespace might not have any labels attached to it. Directly accessing the label map
+		// leads to a crash for such namespaces.
+		// If the label map is nil, create one and use the setter api to attach it to the namespace.
+		nsLabels := namespace.ObjectMeta.GetLabels()
+		if nsLabels == nil {
+			nsLabels = make(map[string]string)
+		}
+		nsLabels[AllowedNamespaceSelectorLabelKey] = cfgAllowedNs
+		namespace.ObjectMeta.SetLabels(nsLabels)
+
+		err = r.Update(ctx, namespace)
+		if err != nil {
+			log.Error(err, "Failed to label namespace", "namespace", cfgAllowedNs)
+			return err
+		}
+		log.Info("Labeled namespace successfully", "namespace", cfgAllowedNs)
+
+		labeledAllowedNsList = append(labeledAllowedNsList, cfgAllowedNs)
+		statusChanged = true
+	}
+	// Sweep the existing namespaces again to unbind any namespace that was not found in the configured list
+	for existingAllowedNs, _ := range existingAllowedNsMap {
```

Comment:
If someone removes from the list, the network policy gets updated. If you remove a ns called nginx-ingress from the allowed ns list, the network policy will remove the label kubeslice.io/namespace: nginx-ingress.

Reply:
yaa but the label on the namespace still remains, this func just removes the label(kubeslice.io/namespace=.) from that ns

Reply:
Then how will this work if the namespace is listed in allowed ns lists of more than one slice? If you remove the label here then other slices would break.